File: server/src/store/sessions.ts
import type { Database } from 'bun:sqlite'
import { randomUUID } from 'node:crypto'

import type { StoredSession, VersionedUpdateResult } from './types'
import { safeJsonParse } from './json'
import { updateVersionedField } from './versionedUpdates'

type DbSessionRow = {
    id: string
    tag: string | null
    namespace: string
    machine_id: string | null
    created_at: number
    updated_at: number
    metadata: string | null
    metadata_version: number
    agent_state: string | null
    agent_state_version: number
    todos: string | null
    todos_updated_at: number | null
    active: number
    active_at: number | null
    seq: number
}

function toStoredSession(row: DbSessionRow): StoredSession {
    return {
        id: row.id,
        tag: row.tag,
        namespace: row.namespace,
        machineId: row.machine_id,
        createdAt: row.created_at,
        updatedAt: row.updated_at,
        metadata: safeJsonParse(row.metadata),
        metadataVersion: row.metadata_version,
        agentState: safeJsonParse(row.agent_state),
        agentStateVersion: row.agent_state_version,
        todos: safeJsonParse(row.todos),
        todosUpdatedAt: row.todos_updated_at,
        active: row.active === 1,
        activeAt: row.active_at,
        seq: row.seq
    }
}

export function getOrCreateSession(
    db: Database,
    id: string | null,
    metadata: unknown,
    agentState: unknown,
    namespace: string
): StoredSession {
    console.log('[SessionStore.getOrCreateSession] Called with:', { id, namespace })

    // Validate UUID format if id is provided
    if (id !== null && !isValidUUID(id)) {
        console.error('[SessionStore.getOrCreateSession] Invalid UUID format:', { id })
        throw new Error(`Invalid session ID format: ${id}`)
    }

    // Generate UUID if not provided (new session)
    const sessionId = id ?? randomUUID()

    // CRITICAL CHANGE: Look up by id (PRIMARY KEY) instead of tag
    const existing = db.prepare(
        'SELECT * FROM sessions WHERE id = ? AND namespace = ? ORDER BY created_at DESC LIMIT 1'
    ).get(sessionId, namespace) as DbSessionRow | undefined

    if (existing) {
        console.log('[SessionStore.getOrCreateSession] Found existing session:', {
            id: existing.id,
            namespace: existing.namespace
        })
        return toStoredSession(existing)
    }

    // CRITICAL ERROR HANDLING: If id was provided but not found, session doesn't exist
    if (id !== null) {
        console.error('[SessionStore.getOrCreateSession] Session not found:', { id, namespace })
        throw new Error(`Session not found: ${id}`)
    }

    // Create new session with generated UUID
    console.log('[SessionStore.getOrCreateSession] Creating NEW session:', {
        id: sessionId,
        namespace
    })

    const now = Date.now()
    const metadataJson = JSON.stringify(metadata)
    const agentStateJson = agentState === null || agentState === undefined ? null : JSON.stringify(agentState)

    db.prepare(`
        INSERT INTO sessions (
            id, tag, namespace, machine_id, created_at, updated_at,
            metadata, metadata_version,
            agent_state, agent_state_version,
            todos, todos_updated_at,
            active, active_at, seq
        ) VALUES (
            @id, @tag, @namespace, NULL, @created_at, @updated_at,
            @metadata, 1,
            @agent_state, 1,
            NULL, NULL,
            0, NULL, 0
        )
    `).run({
        id: sessionId,
        tag: sessionId,  // Keep tag = id for now (no schema migration)
        namespace,
        created_at: now,
        updated_at: now,
        metadata: metadataJson,
        agent_state: agentStateJson
    })

    console.log('[SessionStore.getOrCreateSession] New session created successfully:', { id: sessionId })

    const row = getSession(db, sessionId)
    if (!row) {
        throw new Error('Failed to create session')
    }
    return row
}

// UUID validation helper
function isValidUUID(str: string): boolean {
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i
    return uuidRegex.test(str)
}

export function createSession(
    db: Database,
    id: string,
    namespace: string,
    metadata: unknown,
    agentState: unknown,
    active: boolean = false,
    thinking: boolean = false
): StoredSession {
    if (!isValidUUID(id)) {
        throw new Error(`Invalid session ID format: ${id}`)
    }

    const now = Date.now()
    const metadataJson = JSON.stringify(metadata)
    const agentStateJson = agentState === null || agentState === undefined ? null : JSON.stringify(agentState)

    db.prepare(`
        INSERT INTO sessions (
            id, tag, namespace, machine_id, created_at, updated_at,
            metadata, metadata_version,
            agent_state, agent_state_version,
            todos, todos_updated_at,
            active, active_at, seq
        ) VALUES (
            @id, @tag, @namespace, NULL, @created_at, @updated_at,
            @metadata, 1,
            @agent_state, 1,
            NULL, NULL,
            @active, @active_at, 0
        )
    `).run({
        id,
        tag: id,
        namespace,
        created_at: now,
        updated_at: now,
        metadata: metadataJson,
        agent_state: agentStateJson,
        active: active ? 1 : 0,
        active_at: active ? now : null
    })

    const row = getSession(db, id)
    if (!row) {
        throw new Error('Failed to create session')
    }
    return row
}

export function updateSessionMetadata(
    db: Database,
    id: string,
    metadata: unknown,
    expectedVersion: number,
    namespace: string,
    options?: { touchUpdatedAt?: boolean }
): VersionedUpdateResult<unknown | null> {
    const now = Date.now()
    const touchUpdatedAt = options?.touchUpdatedAt !== false

    return updateVersionedField({
        db,
        table: 'sessions',
        id,
        namespace,
        field: 'metadata',
        versionField: 'metadata_version',
        expectedVersion,
        value: metadata,
        encode: (value) => {
            const json = JSON.stringify(value)
            return json === undefined ? null : json
        },
        decode: safeJsonParse,
        setClauses: [
            'updated_at = CASE WHEN @touch_updated_at = 1 THEN @updated_at ELSE updated_at END',
            'seq = seq + 1'
        ],
        params: {
            updated_at: now,
            touch_updated_at: touchUpdatedAt ? 1 : 0
        }
    })
}

export function updateSessionAgentState(
    db: Database,
    id: string,
    agentState: unknown,
    expectedVersion: number,
    namespace: string
): VersionedUpdateResult<unknown | null> {
    const now = Date.now()
    const normalized = agentState ?? null

    return updateVersionedField({
        db,
        table: 'sessions',
        id,
        namespace,
        field: 'agent_state',
        versionField: 'agent_state_version',
        expectedVersion,
        value: normalized,
        encode: (value) => (value === null ? null : JSON.stringify(value)),
        decode: safeJsonParse,
        setClauses: ['updated_at = @updated_at', 'seq = seq + 1'],
        params: { updated_at: now }
    })
}

export function setSessionTodos(
    db: Database,
    id: string,
    todos: unknown,
    todosUpdatedAt: number,
    namespace: string
): boolean {
    try {
        const json = todos === null || todos === undefined ? null : JSON.stringify(todos)
        const result = db.prepare(`
            UPDATE sessions
            SET todos = @todos,
                todos_updated_at = @todos_updated_at,
                updated_at = CASE WHEN updated_at > @updated_at THEN updated_at ELSE @updated_at END,
                seq = seq + 1
            WHERE id = @id
              AND namespace = @namespace
              AND (todos_updated_at IS NULL OR todos_updated_at < @todos_updated_at)
        `).run({
            id,
            todos: json,
            todos_updated_at: todosUpdatedAt,
            updated_at: todosUpdatedAt,
            namespace
        })

        return result.changes === 1
    } catch {
        return false
    }
}

export function getSession(db: Database, id: string): StoredSession | null {
    const row = db.prepare('SELECT * FROM sessions WHERE id = ?').get(id) as DbSessionRow | undefined
    return row ? toStoredSession(row) : null
}

export function getSessionByNamespace(db: Database, id: string, namespace: string): StoredSession | null {
    const row = db.prepare(
        'SELECT * FROM sessions WHERE id = ? AND namespace = ?'
    ).get(id, namespace) as DbSessionRow | undefined
    return row ? toStoredSession(row) : null
}

export function getSessions(db: Database): StoredSession[] {
    const rows = db.prepare('SELECT * FROM sessions ORDER BY updated_at DESC').all() as DbSessionRow[]
    return rows.map(toStoredSession)
}

export function getSessionsByNamespace(db: Database, namespace: string): StoredSession[] {
    const rows = db.prepare(
        'SELECT * FROM sessions WHERE namespace = ? ORDER BY updated_at DESC'
    ).all(namespace) as DbSessionRow[]
    return rows.map(toStoredSession)
}

export function deleteSession(db: Database, id: string, namespace: string): boolean {
    const result = db.prepare(
        'DELETE FROM sessions WHERE id = ? AND namespace = ?'
    ).run(id, namespace)
    return result.changes > 0
}

File: hub_go/internal/store/sessions.go
package store

import (
	"database/sql"
	"encoding/json"
	"time"
)

type Session struct {
	ID                   string
	Namespace            string
	Seq                  int64
	CreatedAt            int64
	UpdatedAt            int64
	Active               bool
	ActiveAt             int64
	Metadata             map[string]any
	MetadataVersion      int64
	AgentState           any
	AgentStateVersion    int64
	Todos                any
	TodosUpdatedAt       int64
	PermissionMode       string
	ModelMode            string
	Thinking             bool
	ThinkingAt           int64
	TodoProgress         any
	PendingRequestsCount int
	Tag                  string
}

type UpdateResult[T any] struct {
	Result  string
	Version int64
	Value   T
}

func (s *Store) ListSessions(namespace string) []Session {
	rows, err := s.DB.Query(
		`SELECT id, namespace, seq, created_at, updated_at, active, active_at, metadata, metadata_version, agent_state, agent_state_version, todos, todos_updated_at, permission_mode, model_mode, thinking, thinking_at, tag
         FROM sessions WHERE namespace = ? ORDER BY updated_at DESC`,
		namespace,
	)
	if err != nil {
		return []Session{}
	}
	defer rows.Close()

	var sessions []Session
	for rows.Next() {
		var session Session
		var activeInt int
		var thinkingInt int
		var metadataRaw sql.NullString
		var agentStateRaw sql.NullString
		var todosRaw sql.NullString
		var todosUpdatedAt sql.NullInt64
		var permissionMode sql.NullString
		var modelMode sql.NullString
		var tag sql.NullString
		if err := rows.Scan(
			&session.ID,
			&session.Namespace,
			&session.Seq,
			&session.CreatedAt,
			&session.UpdatedAt,
			&activeInt,
			&session.ActiveAt,
			&metadataRaw,
			&session.MetadataVersion,
			&agentStateRaw,
			&session.AgentStateVersion,
			&todosRaw,
			&todosUpdatedAt,
			&permissionMode,
			&modelMode,
			&thinkingInt,
			&session.ThinkingAt,
			&tag,
		); err != nil {
			continue
		}
		session.Active = activeInt == 1
		session.Thinking = thinkingInt == 1
		session.Metadata = decodeJSONMap(metadataRaw)
		session.AgentState = decodeJSONValue(agentStateRaw)
		session.Todos = decodeJSONValue(todosRaw)
		if todosUpdatedAt.Valid {
			session.TodosUpdatedAt = todosUpdatedAt.Int64
		}
		if permissionMode.Valid {
			session.PermissionMode = permissionMode.String
		}
		if modelMode.Valid {
			session.ModelMode = modelMode.String
		}
		if tag.Valid {
			session.Tag = tag.String
		}
		sessions = append(sessions, session)
	}
	return sessions
}

func (s *Store) GetSession(namespace string, id string) (*Session, error) {
	row := s.DB.QueryRow(
		`SELECT id, namespace, seq, created_at, updated_at, active, active_at, metadata, metadata_version, agent_state, agent_state_version, todos, todos_updated_at, permission_mode, model_mode, thinking, thinking_at, tag
         FROM sessions WHERE id = ? AND namespace = ? LIMIT 1`,
		id,
		namespace,
	)

	var session Session
	var activeInt int
	var thinkingInt int
	var metadataRaw sql.NullString
	var agentStateRaw sql.NullString
	var todosRaw sql.NullString
	var todosUpdatedAt sql.NullInt64
	var permissionMode sql.NullString
	var modelMode sql.NullString
	var tag sql.NullString
	if err := row.Scan(
		&session.ID,
		&session.Namespace,
		&session.Seq,
		&session.CreatedAt,
		&session.UpdatedAt,
		&activeInt,
		&session.ActiveAt,
		&metadataRaw,
		&session.MetadataVersion,
		&agentStateRaw,
		&session.AgentStateVersion,
		&todosRaw,
		&todosUpdatedAt,
		&permissionMode,
		&modelMode,
		&thinkingInt,
		&session.ThinkingAt,
		&tag,
	); err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}
	if tag.Valid {
		session.Tag = tag.String
	}
	session.Active = activeInt == 1
	session.Thinking = thinkingInt == 1
	session.Metadata = decodeJSONMap(metadataRaw)
	session.AgentState = decodeJSONValue(agentStateRaw)
	session.Todos = decodeJSONValue(todosRaw)
	if todosUpdatedAt.Valid {
		session.TodosUpdatedAt = todosUpdatedAt.Int64
	}
	if permissionMode.Valid {
		session.PermissionMode = permissionMode.String
	}
	if modelMode.Valid {
		session.ModelMode = modelMode.String
	}
	return &session, nil
}

func (s *Store) SessionExists(id string) bool {
	row := s.DB.QueryRow("SELECT 1 FROM sessions WHERE id = ? LIMIT 1", id)
	var value int
	if err := row.Scan(&value); err != nil {
		return false
	}
	return value == 1
}

func (s *Store) CreateSession(namespace string, metadata map[string]any, agentState any) (*Session, error) {
	return s.CreateSessionWithID(namespace, newUUID(), metadata, agentState)
}

func (s *Store) CreateSessionWithID(namespace string, id string, metadata map[string]any, agentState any) (*Session, error) {
	if id == "" {
		id = newUUID()
	}
	now := time.Now().UnixMilli()
	metadataRaw, _ := json.Marshal(metadata)
	agentStateRaw, _ := json.Marshal(agentState)
	_, err := s.DB.Exec(
		`INSERT OR REPLACE INTO sessions (
            id, namespace, seq, created_at, updated_at, active, active_at, metadata, metadata_version, agent_state, agent_state_version, todos, todos_updated_at, permission_mode, model_mode, thinking, thinking_at, tag
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		id,
		namespace,
		0,
		now,
		now,
		0,
		now,
		string(metadataRaw),
		1,
		string(agentStateRaw),
		1,
		nil,
		nil,
		nil,
		nil,
		0,
		0,
		"",
	)
	if err != nil {
		return nil, err
	}
	return s.GetSession(namespace, id)
}

func (s *Store) UpdateSession(namespace string, session *Session) error {
	if session == nil {
		return nil
	}
	metadataRaw, _ := json.Marshal(session.Metadata)
	agentStateRaw, _ := json.Marshal(session.AgentState)
	todosRaw, _ := json.Marshal(session.Todos)
	active := 0
	if session.Active {
		active = 1
	}
	thinking := 0
	if session.Thinking {
		thinking = 1
	}
	_, err := s.DB.Exec(
		`UPDATE sessions SET
            seq = ?, updated_at = ?, active = ?, active_at = ?, metadata = ?, metadata_version = ?, agent_state = ?, agent_state_version = ?, todos = ?, todos_updated_at = ?, permission_mode = ?, model_mode = ?, thinking = ?, thinking_at = ?, tag = ?
         WHERE id = ? AND namespace = ?`,
		session.Seq,
		session.UpdatedAt,
		active,
		session.ActiveAt,
		string(metadataRaw),
		session.MetadataVersion,
		string(agentStateRaw),
		session.AgentStateVersion,
		string(todosRaw),
		nullInt64(session.TodosUpdatedAt),
		nullString(session.PermissionMode),
		nullString(session.ModelMode),
		thinking,
		session.ThinkingAt,
		session.Tag,
		session.ID,
		namespace,
	)
	return err
}

func (s *Store) UpdateSessionMetadata(namespace string, id string, metadata map[string]any, expectedVersion int64) (UpdateResult[map[string]any], error) {
	session, err := s.GetSession(namespace, id)
	if err != nil {
		return UpdateResult[map[string]any]{Result: "error"}, err
	}
	if session == nil {
		return UpdateResult[map[string]any]{Result: "error"}, nil
	}
	if expectedVersion != session.MetadataVersion {
		return UpdateResult[map[string]any]{Result: "version-mismatch", Version: session.MetadataVersion, Value: session.Metadata}, nil
	}

	session.MetadataVersion++
	session.Metadata = metadata
	session.UpdatedAt = time.Now().UnixMilli()
	if err := s.UpdateSession(namespace, session); err != nil {
		return UpdateResult[map[string]any]{Result: "error"}, err
	}
	return UpdateResult[map[string]any]{Result: "success", Version: session.MetadataVersion, Value: session.Metadata}, nil
}

func (s *Store) UpdateSessionAgentState(namespace string, id string, agentState any, expectedVersion int64) (UpdateResult[any], error) {
	session, err := s.GetSession(namespace, id)
	if err != nil {
		return UpdateResult[any]{Result: "error"}, err
	}
	if session == nil {
		return UpdateResult[any]{Result: "error"}, nil
	}
	if expectedVersion != session.AgentStateVersion {
		return UpdateResult[any]{Result: "version-mismatch", Version: session.AgentStateVersion, Value: session.AgentState}, nil
	}

	session.AgentStateVersion++
	session.AgentState = agentState
	session.UpdatedAt = time.Now().UnixMilli()
	if err := s.UpdateSession(namespace, session); err != nil {
		return UpdateResult[any]{Result: "error"}, err
	}
	return UpdateResult[any]{Result: "success", Version: session.AgentStateVersion, Value: session.AgentState}, nil
}

func (s *Store) SetSessionTodos(namespace string, id string, todos any, todosUpdatedAt int64) bool {
	session, err := s.GetSession(namespace, id)
	if err != nil || session == nil {
		return false
	}
	if session.TodosUpdatedAt != 0 && session.TodosUpdatedAt >= todosUpdatedAt {
		return false
	}
	session.Todos = todos
	session.TodosUpdatedAt = todosUpdatedAt
	session.UpdatedAt = todosUpdatedAt
	if err := s.UpdateSession(namespace, session); err != nil {
		return false
	}
	return true
}

func (s *Store) DeleteSession(namespace string, id string) bool {
	result, err := s.DB.Exec("DELETE FROM sessions WHERE id = ? AND namespace = ?", id, namespace)
	if err != nil {
		return false
	}
	rows, _ := result.RowsAffected()
	return rows > 0
}

func decodeJSONMap(raw sql.NullString) map[string]any {
	if !raw.Valid || raw.String == "" {
		return nil
	}
	var value map[string]any
	if err := json.Unmarshal([]byte(raw.String), &value); err != nil {
		return nil
	}
	return value
}

func decodeJSONValue(raw sql.NullString) any {
	if !raw.Valid || raw.String == "" {
		return nil
	}
	var value any
	if err := json.Unmarshal([]byte(raw.String), &value); err != nil {
		return nil
	}
	return value
}

func nullString(value string) any {
	if value == "" {
		return nil
	}
	return value
}

func nullInt64(value int64) any {
	if value == 0 {
		return nil
	}
	return value
}
